sched/scheduler/scheduler.go

```diff
@@ -0,0 +1,121 @@
+package scheduler
+
+import (
+	"github.com/scootdev/scoot/saga"
+	"github.com/scootdev/scoot/sched"
+	cm "github.com/scootdev/scoot/sched/clustermembership"
+	dist "github.com/scootdev/scoot/sched/distributor"
+	"sync"
+)
+
+type scheduler struct {
+	cluster     cm.DynamicCluster
+	saga        saga.Saga
+	distributor *dist.PoolDistributor
+	started     bool           // tracks if Scheduler has been Initialized
+	wg          sync.WaitGroup // used to track jobs in progress
+}
+
+func NewScheduler(cluster cm.DynamicCluster, clusterState cm.DynamicClusterState, saga saga.Saga) *scheduler {
+	return &scheduler{
+		cluster:     cluster,
+		saga:        saga,
+		distributor: dist.NewDynamicPoolDistributor(clusterState),
+		started:     false,
+	}
+}
+
+// Blocks until all scheduled jobs are compeleted
+func (s *scheduler) BlockUnitlAllJobsCompleted() {
+	s.wg.Wait()
+}
+
+// Starts the scheduler, must be called before any other
+// methods on the scheduler can be called
+func (s *scheduler) Start() {
+
+	// Recover form SagaLog Any In Process Tasks
+	// Return only once all those have been scheduled
+
+	s.started = true
+}
+
+// Schedule a job, returns once the job has been successfully
+// scheduled, nodes reserved & durably started Saga,
+// Returns an error if scheduling was unsuccessful
+func (s *scheduler) ScheduleJob(job sched.Job) error {
+
+	if !s.started {
+		return newUninitializedSchedError()
+	}
+
+	// Log StartSaga Message
+	// TODO: need to serialize job into binary and pass in here
+	// so we can recover the job in case of failure
+	sagaState, err := s.saga.StartSaga(job.Id, nil)
+
+	// If we succssfully started the Saga, ProcssJob
+	if err == nil {
+
+		// Reserve Nodes to Schedule Job On
```

Comment:
This is too removed from real issues, and no matter the approach, this will be 'easy'.

For a motivating case, how about a job with 1 long running task and a bunch of short lived tasks. We'll be tying up multiple nodes doing nothing because we can't free our pool until everything is done. In that case, I would want to pass the pool down to job runner. The only issue with that is it becomes harder to provide back pressure.